phoenicis-apps/src/main/java/org/phoenicis/apps/GitApplicationsSource.java

```diff
@@ -18,45 +18,109 @@
+	private List<CategoryDTO> cache;
+
+	private static final String cacheDirectoryPath = System.getProperty("java.io.tmpdir");
+
+	GitApplicationsSource(String gitRepositoryURL, LocalApplicationsSource.Factory localAppsSourceFactory) {
+		this.gitRepositoryURL = gitRepositoryURL;
+		this.localAppsSourceFactory = localAppsSourceFactory;
+	}
+
+	@Override
+	public synchronized List<CategoryDTO> fetchInstallableApplications() {
+		if (cache != null) {
+			return cache;
+		}
+
+		File gitRepositoryLocation = new File(
+				GitApplicationsSource.cacheDirectoryPath + "/git" + this.gitRepositoryURL.hashCode());
+
+		LOGGER.info(String.format("Git-repository '%s' will be saved in '%s'", this.gitRepositoryURL,
+				gitRepositoryLocation.getAbsolutePath()));
+
+		/*
+		 * This value is true if a folder for the git repository exists,
+		 * otherwise it is false
+		 */
+		boolean folderExists = true;
+
+		// check that the repository folder exists
+		if (!gitRepositoryLocation.exists()) {
+			folderExists = false;
+
+			LOGGER.info(String.format("Creating new folder '%s' for git-repository '%s'",
+					gitRepositoryLocation.getAbsolutePath(), this.gitRepositoryURL));
+
+			if (!gitRepositoryLocation.mkdir()) {
+				LOGGER.error(String.format("Couldn't create folder for git repository '%s' at '%s'",
+						this.gitRepositoryURL, gitRepositoryLocation.getAbsolutePath()));
+
+				return Collections.emptyList();
+			}
+		}
+
+		try {
+			Git gitRepository = null;
+
+			/*
+			 * if the repository folder previously didn't exist, clone the
+			 * repository now
+			 */
+			if (!folderExists) {
+				LOGGER.info(String.format("Cloning git-repository '%s' to '%s'", this.gitRepositoryURL, gitRepositoryLocation.getAbsolutePath()));
+
+				gitRepository = Git.cloneRepository().setURI(this.gitRepositoryURL).setDirectory(gitRepositoryLocation)
+						.call();
+			}
+			/*
+			 * otherwise open the folder and pull the newest updates from the
+			 * repository
+			 */
+			else {
+				LOGGER.info(String.format("Opening git-repository '%s' at '%s'", this.gitRepositoryURL,
+						gitRepositoryLocation.getAbsolutePath()));
+
+				gitRepository = Git.open(gitRepositoryLocation);
+
+				LOGGER.info(String.format("Pulling new commits from git-repository '%s' to '%s'", this.gitRepositoryURL,
+						gitRepositoryLocation.getAbsolutePath()));
+
+				gitRepository.pull().call();
+			}
+
+			// close repository to free resources
+			gitRepository.close();
+
+			this.cache = localAppsSourceFactory.createInstance(gitRepositoryLocation.getAbsolutePath())
+					.fetchInstallableApplications();
+		} catch (RepositoryNotFoundException | GitAPIException e) {
+			LOGGER.error(String.format("Folder '%s' is no git-repository", gitRepositoryLocation.getAbsolutePath()), e);
+
+			this.cache = Collections.emptyList();
+		} catch (IOException e) {
+			LOGGER.error(String.format("An unknown error occured.", e));
 
+			this.cache = Collections.emptyList();
+		}
 
+		return this.cache;
```

Comment:
cache might be null here. It should be an empty list if something failed (I don't think it will work if you try to clone again later).